babel-macro/src/macro.ts

```diff
@@ -24,15 +24,48 @@ interface MacrosBabel {
   types: typeof babelTypes;
 }
 
-function findExportedMethods(ast: babelTypes.File, { types: t }: MacrosBabel): string[] {
+function getFunctionName(e: babelTypes.FunctionDeclaration, t: typeof babelTypes): string | undefined {
+  return t.isIdentifier(e.id) ? e.id.name : undefined;
+}
+
+function verifyExportedMember(ast: babelTypes.File, t: typeof babelTypes, funcName: string): void {
+  const body = ast.program.body;
+  const isFunctionExported = body.some((e) =>
+    t.isExpressionStatement(e) &&
+    t.isAssignmentExpression(e.expression) &&
+    e.expression.operator === '=' &&
+    t.isMemberExpression(e.expression.left) &&
+    t.isIdentifier(e.expression.left.object) &&
+    e.expression.left.object.name === 'exports' &&
+    t.isIdentifier(e.expression.left.property) &&
+    e.expression.left.property.name === funcName &&
+    t.isIdentifier(e.expression.right) &&
+    e.expression.right.name === funcName
+  );
+  if (!isFunctionExported) {
+    throw new Error(`${funcName} has @expose decorator but it is not exported`);
+  }
+}
+
+function findExportedMethods(ast: babelTypes.File, { types: t }: MacrosBabel, importedNames: string[]): string[] {
   const body = ast.program.body;
   // support ExportDefaultDeclaration (ExportAllDeclaration too for re-exporting ?)
-  const namedExports = body.filter((e): e is babelTypes.ExportNamedDeclaration => t.isExportNamedDeclaration(e));
-  const exposedExports = namedExports.filter(
-    (e) => e.leadingComments && e.leadingComments.some((comment) => /@expose/.test(comment.value)));
-  const functions = exposedExports.map((e) => e.declaration).filter(
-    (e): e is babelTypes.FunctionDeclaration => t.isFunctionDeclaration(e));
-  return functions.filter((f) => !!f.id).map((f) => f.id!.name);
+  const exposedStatements = body.filter((e) => e.leadingComments && e.leadingComments.some((comment) => /@expose/.test(comment.value)));
+  return exposedStatements.reduce((ret: string[], e) => {
+    if (t.isFunctionDeclaration(e)) {
+      const funcName = getFunctionName(e, t);
+      if (funcName && importedNames.includes(funcName)) {
```

Comment:
why limit this only to imported names (`importedNames.includes(funcName)`)? So we can throw during import?
if so, why doesn't the function name in line 62 gets the same treatment?

Reply:
given this `backend` file:
```js
// @expose
export function foo() {
}

// @expose
function internalFunctionWithObsoleteExposeComment() {
}
```

And N frontend files like this:
```js
import { foo } from '../backend/myFile'

...
```

I don't want all the frontend files compilations to fail because of an unrelated compilation error in a different file that would not cause anything wrong in the front-end code.
At most, I want this error only in the backend file compilation.

When importing `foo` I do want this error because this function is actually called from the frontend code which I'm currently compiling, so it wouldn't work if the function is not exported and exposed